zkevm-circuits/src/evm_circuit.rs

```diff
@@ -1 +1,930 @@
+    }
+}
+
+// TODO: Integrate with evm_word
+#[derive(Clone, Debug)]
+struct Word<F> {
+    // random linear combination expression of cells
+    expression: Expression<F>,
+    // inner cells for syntesis
+    cells: [Cell<F>; 32],
+}
+
+impl<F: FieldExt> Word<F> {
+    fn new(cells: &[Cell<F>], r: F) -> Self {
+        let r = Expression::Constant(r);
+        Self {
+            expression: cells
+                .iter()
+                .rev()
+                .fold(Expression::Constant(F::zero()), |acc, byte| {
+                    acc * r.clone() + byte.expr()
+                }),
+            cells: cells.to_owned().try_into().unwrap(),
+        }
+    }
+
+    fn expr(&self) -> Expression<F> {
+        self.expression.clone()
+    }
+
+    fn assign(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        word: Option<[u8; 32]>,
+    ) -> Result<Vec<circuit::Cell>, Error> {
+        word.map_or(Err(Error::SynthesisError), |word| {
+            self.cells
+                .iter()
+                .zip(word.iter())
+                .map(|(cell, byte)| {
+                    cell.assign(region, offset, Some(F::from_u64(*byte as u64)))
+                })
+                .collect()
+        })
+    }
+}
+
+// TODO: Should we maintain these state in circuit or we prepare all of them
+// when parsing debug trace?
+#[derive(Debug)]
+pub(crate) struct CoreStateInstance {
+    is_executing: bool,
+    global_counter: usize,
+    call_id: usize,
+    program_counter: usize,
+    stack_pointer: usize,
+    gas_counter: usize,
+}
+
+impl CoreStateInstance {
+    fn new(call_id: usize) -> Self {
+        Self {
+            is_executing: false,
+            global_counter: 1,
+            call_id,
+            program_counter: 0,
+            stack_pointer: 1024,
+            gas_counter: 0,
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+enum Case {
+    Success,
+    // out of gas
+    OutOfGas,
+    // call depth exceeded 1024
+    DepthOverflow,
+    // insufficient balance for transfer
+    InsufficientBalance,
+    // contract address collision
+    ContractAddressCollision,
+    // execution reverted
+    Reverted,
+    // max code size exceeded
+    MaxCodeSizeExceeded,
+    // invalid jump destination
+    InvalidJump,
+    // static call protection
+    WriteProtection,
+    // return data out of bound
+    ReturnDataOutOfBounds,
+    // contract must not begin with 0xef due to EIP #3541 EVM Object Format
+    // (EOF)
+    InvalidBeginningCode,
+    // stack underflow
+    StackUnderflow,
+    // stack overflow
+    StackOverflow,
+    // invalid opcode
+    InvalidCode,
+}
+
+// TODO: use ExecutionStep from bus_mapping
+pub(crate) struct ExecutionStep {
+    opcode: u8,
+    case: Case,
+    values: Vec<[u8; 32]>,
+}
+
+// TODO: use Operation from bus_mapping
+pub(crate) struct Operation<F> {
+    gc: usize,
+    target: Target,
+    is_write: bool,
+    values: [F; 4],
+}
```

Comment:
Should we use in this PR the real types? Or you want to wait until we parse all the info from geth? As long as it doesn't affect the Opcode, I don't really care. But we should file issues for that.

Reply:
I tried but found I have to change a lot on `bus-mapping` side (export interface, add methods, etc...), which could take me some more days to finish this. So I plan to use a simple struct for this PR, and file an issue for what `evm-circuit` would need from `bus-mapping`.

Reply:
Perfect! Let's file an issue bofore merging then! :smile:

Reply:
Mentioned in #61